### PR TITLE
Jonathan/appeals 36689

### PR DESCRIPTION
**Pull request by jtsangVA**
Resolves [APPEALS-36689](https://jira.devops.va.gov/browse/APPEALS-36689)

# Description
As a developer, I need to build out the DecisionReviewCreated service class that will handle logic around creating intake records. The class will be leveraged by the DecisionReviewCreated controller when our API is called.

## Acceptance Criteria
- [x] Code compiles correctly
- [x] Build out the Create controller action that will call on this service class when the api/events/v1/decision_review_created route is called.
- [x] Service class will check if event_id already exists within events table (and was completed successfully) OR if claim_id is already stored within RedisMutex
- [x] If event does not exist, RedisMutex gem will create lock based off of claim_id and have a with a 60 second expiry
Service class will be within the app/services/events directory and be named decision_review_created
- [x] Event instance will be persisted to the database.
- [x] Create RSPEC scenarios for all above AC.

## Testing Plan
1. Test Plan: https://jira.devops.va.gov/browse/APPEALS-39790
2. Execution: https://jira.devops.va.gov/browse/APPEALS-39797
3. X-ray: https://jira.devops.va.gov/secure/XrayExecuteTest!default.jspa?testExecIssueKey=APPEALS-39797&testIssueKey=APPEALS-39790

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Backend

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [x] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Logging
- [x] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

**Review by Jruuuu — COMMENTED** [review]
LGTM Well done!

File: app/services/events/decision_review_created.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        event = find_or_create_event(consumer_event_id)
+
+        ActiveRecord::Base.transaction do
+          # TODO: backfill models as needed, set Event.completed_at when finished
+          # event.update!(completed_at: Time.now, error: nil)
+        end
+      end
+      true
+    rescue RedisMutex::LockError
+      Rails.logger.error("Failed to acquire lock for Claim ID: #{reference_id}! This Event is being"\
```
Comment (nkutub) [inline app/services/events/decision_review_created.rb:29]: Please see other PR , I think we are assuming the wrong reasons an error will be raised here.

File: app/services/events/decision_review_created.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    # Check if there's already a CF Event that references that Appeals-Consumer EventID and
+    # was successfully completed
+    def event_exists_and_is_completed?(consumer_event_id)
+      Event.where(reference_id: consumer_event_id).where.not(completed_at: nil).exists?
+    end
+
+    # Check if there's already a CF Event that references that Appeals-Consumer EventID
+    # We will update the existing Event instead of creating a new one
+    def find_or_create_event(consumer_event_id)
+      Event.find_by(reference_id: consumer_event_id) || DecisionReviewCreatedEvent.create!(reference_id: consumer_event_id)
```
Comment (nkutub) [inline app/services/events/decision_review_created.rb:46]: you might want to look at find_or_create_by to simplify this logic

File: app/services/events/decision_review_created.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  # Default options for RedisMutex#with_lock
+  # :block  => 1    # Specify in seconds how long you want to wait for the lock to be released.
+  #                 # Specify 0 if you need non-blocking sematics and return false immediately. (default: 1)
+  # :sleep  => 0.1  # Specify in seconds how long the polling interval should be when :block is given.
+  #                 # It is NOT recommended to go below 0.01. (default: 0.1)
+  # :expire => 10   # Specify in seconds when the lock should be considered stale when something went wrong
+  #                 # with the one who held the lock and failed to unlock. (default: 10)
+
+  class << self
+    def create(consumer_event_id, reference_id)
```
Comment (nkutub) [inline app/services/events/decision_review_created.rb:14]: typically if a method could raise an error it should be defined with ! at end but I'm not gonna be that picky about it

File: app/services/events/decision_review_created.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      return if event_exists_and_is_completed?(consumer_event_id)
+
+      redis = Redis.new(url: Rails.application.secrets.redis_url_cache)
+      # Use the consumer_event_id to retrieve/create the Event object
+      event = find_or_create_event(consumer_event_id)
```
Comment (nkutub) [inline app/services/events/decision_review_created.rb:19]: I don't think we should create the even before we check redis , I think this should be moved down

Comment (nkutub) [reply app/services/events/decision_review_created.rb:19]: I think this should be in safe we should move the creation of the event in the RedisMutex block to be safe

**Review by nkutub — APPROVED** [review]

**Pull request merged**